src/main/java/de/zalando/aruha/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -0,0 +1,366 @@
+package de.zalando.aruha.nakadi.service.subscription.state;
+
+import de.zalando.aruha.nakadi.service.EventStream;
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZKSubscription;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class StreamingState extends State {
+    private ZKSubscription topologyChangeSubscription;
+    private Consumer<String, String> kafkaConsumer;
+    private final Map<Partition.PartitionKey, PartitionData> offsets = new HashMap<>();
+    // Maps partition barrier when releasing must be completed or stream will be closed.
+    // The reasons for that if there are two partitions (p0, p1) and p0 is reassigned, if p1 is working
+    // correctly, and p0 is not receiving any updates - reassignment won't complete.
+    private final Map<Partition.PartitionKey, Long> releasingPartitions = new HashMap<>();
+    private boolean pollPaused = false;
+    private long lastCommitMillis;
+    private long commitedEvents = 0;
+    private long sentEvents = 0;
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamingState.class);
+
+    @Override
+    public void onEnter() {
+        // Create kafka consumer
+        this.kafkaConsumer = getKafka().createKafkaConsumer();
+        // Subscribe for topology changes.
+        this.topologyChangeSubscription = getZk().subscribeForTopologyChanges(() -> addTask(this::topologyChanged));
+        // and call directly
+        reactNoTopologyChange();
+        addTask(this::pollDataFromKafka);
+        scheduleTask(this::checkBatchTimeouts, getParameters().batchTimeoutMillis, TimeUnit.MILLISECONDS);
+
+        if (null != getParameters().streamTimeoutMillis) {
+            scheduleTask(this::shutdownGracefully, getParameters().streamTimeoutMillis, TimeUnit.MILLISECONDS);
+        }
+
+        this.lastCommitMillis = System.currentTimeMillis();
+        scheduleTask(this::checkCommitTimeout, getParameters().commitTimeoutMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private void checkCommitTimeout() {
+        if (!isCurrent()) {
```

Comment:
This task is scheduled once and than it's just rescheduled. If Exception occurred during offsets refresh or in some other case - the state will probably switch to some other state. For example - there are 2 tasks in queue: `p1` and `checkCommitTimeout`. `checkCommitTimeout` is deferred and will be executed after 1 second. p1 generates exception, which causes change in tasks queue, now its `switchState` and `checkCommitTimeout`. `switchState` will switch state once executed, after that there will be only one task - `checkCommitTimeout`. But it's strange to execute it, because no one cares about commit timeouts anymore.
There are basically 3 approaches for that:
 - Remember current state while scheduling task (and this can be easily done)
 - Make StreamingContext act as a proxy for states (complex, but it's standard approach for state machines)
 - Ignore processing in each case where it's needed. 

I like the first approach, but it was simpler to debug and understand what is going on with the last one.